src/Features/Core/Portable/UnusedReferences/UnusedReferencesService.cs

```diff
@@ -0,0 +1,132 @@
+            var remainingReferences = references.Except(toplevelUsedReferences);
+            var remainingUsedAssemblyReferences = usedAssemblyReferences.Except(toplevelUsedAssemblyReferences);
+            var remainingUsedAssemblyLookup = remainingUsedAssemblyReferences.ToImmutableHashSet();
+
+            // Determine which transtive references have compilation assemblies in the set of remaining used assemblies.
+            foreach (var reference in remainingReferences)
+            {
+                var allCompilationAssemblies = reference.GetAllCompilationAssemblies();
+                if (allCompilationAssemblies.IsEmpty)
+                {
+                    // Analyzer packages won't contribute any assemblies to the compilation but will
+                    // show up as a reference. We will consider them used.
+                    continue;
+                }
+
+                if (!allCompilationAssemblies.Any(assembly => remainingUsedAssemblyLookup.Contains(assembly)))
+                {
+                    // References that do not contribute a used compilation assembly are considered unsused.
+                    unusedReferences.Add(reference);
+                    continue;
+                }
+
+                remainingUsedAssemblyReferences = remainingUsedAssemblyReferences.Except(allCompilationAssemblies);
+                remainingUsedAssemblyLookup = remainingUsedAssemblyReferences.ToImmutableHashSet();
+            }
+
+            return remainingUsedAssemblyReferences;
+        }
+
+        public async Task<Project> UpdateReferencesAsync(Project project, ImmutableArray<ReferenceUpdate> referenceUpdates, CancellationToken cancellationToken)
+        {
+            var referenceCleanupService = project.Solution.Workspace.Services.GetRequiredService<IReferenceCleanupService>();
+
+            // Remove unused usings
+
+            foreach (var referenceUpdate in referenceUpdates)
+            {
+                if (referenceUpdate.Action == UpdateAction.None)
+                {
+                    referenceUpdate.Action = referenceUpdate.ReferenceInfo.TreatAsUsed
+                        ? UpdateAction.TreatAsUnused
+                        : UpdateAction.None;
+                }
+                else if (referenceUpdate.Action == UpdateAction.TreatAsUsed)
+                {
+                    referenceUpdate.Action = referenceUpdate.ReferenceInfo.TreatAsUsed
+                        ? UpdateAction.None
+                        : UpdateAction.TreatAsUsed;
+                }
+
+                if (referenceUpdate.Action == UpdateAction.None)
+                {
+                    continue;
+                }
+
+                await referenceCleanupService.TryUpdateReferenceAsync(
+                    project.FilePath,
+                    referenceUpdate,
+                    cancellationToken).ConfigureAwait(true);
+            }
+
+            return project;
```

Comment:
Why does this return the project that was passed in, when it doesn't modify the project at all?

(or does it, somehow? if so, I think thats even more confusing 😛)

Reply:
I updated this bit to return the project from the CurrentSolution after applying the updates.